walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,144 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.ListIterator;
 
+/**
+ * Returns a range of possible meeting times given events of individuals, the
+ * length of the meeting, and the people involved.
+ */
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  // Comparator using the Java 8 lambda syntax.
+  private final Comparator<Event> ORDER_END =
+      (Event a, Event b) -> Long.compare(a.getWhen().end(), b.getWhen().end());
+
+  public Collection<TimeRange> query(Collection<Event> events,
+                                     MeetingRequest request) {
+    /* Algorithm overview:
+    Remove people from the event list who are in the list but not in the
+    meeting.
+    Sort all of events in descending order by end time.
+    Find all events that intersect with the current event and create event
+    encompassing them.
+    Continue algorithm on next event after events used.
+    Grab the start
+    time for this current event and traverse through array all events that have
+    an end time greater than this start time. The time slot that can not be used
+    is then the end time of the current iteration and the minimum start time of
+    those other events. Now take the event that after the last event that was
+    discovered still within the constraints and repeat from there.
+
+    To handle the optional attendees case we essentially run the algorithm
+    twice, thus giving us the same overall complexity and general structure.
+
+    Complexity:
+    Such an approach will take O(nlogn) time to sort and O(n) time to iterate
+    where n is the number of meetings and m is the number of people in the
+    new meeting. The removal will take O(mn) time giving us a complexity of O(mn
+    + nlogn).
+    */
+
+    // Store the times when we cannot meet, essentially take the inverse of this
+    // to determine the times we can meet.
+    Collection<String> meetingAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<TimeRange> toReturn = new ArrayList<>();
```

Comment:
The name "toReturn" is not meaningful enough, maybe call it avaliableTimeSlot or whatever.